### ISSUE TITLE
I have found some errors in the use of the indices returned by HungarianMatcher

### ISSUE THREAD

**Issue by FT115** [orrzohar/PROB#55]
Dear author, 
First of all, thank you for your great work. 
As described in the title, if I understand correctly, the indices returned by HungarianMatcher represent the match between the query and the target in the label. However, during the ExemplarSelection phase after completing training and evaluation, the code does not seem to be like this. The 585th line of probdeformable_detr. py seems to be incorrect, which contradicts the 334th line. The `prob[i, indices[i][0], targets[i]['labels']` in line 585 do not seem to use matching indices, which may cause query and target matching errors.

**Comment by orrzohar** [comment]
Hi @FT115,

I look into this, but I think that the lines are different as in line 334, you want the target index.

Meanwhile in line 585, we are using the _source_ index, as we want the **score** of that prediction. 

The `prob` array is indexed by the model predictions, rather than the GT labels. Unlike this, in line 334, `t["labels"]` is indexed by the GT indices. This is why in one, we want the source index and in the other the GT index. I will, however, run the code in debug to make sure this is the case.

Best,
Orr

**Comment by FT115** [comment]
Thanks for your prompt reply, but it seems to me that here prob is used as value and the corresponding key is the GT label, I think the code at line 585 should be: `image_sorted_scores[''.join([chr(int(c)) for c in targets[i]['org_image_id']])] = {'labels':targets[i]['labels'][indices[i][1]].cpu().numpy(),"scores": prob[i,indices[i][0],targets[i]['labels'][indices[i][1]]].detach().cpu().numpy()}`

**Comment by FT115** [comment]
Thanks for your prompt reply, but it seems to me that here `prob` is used as value in dictionary, and another value is the GT label. So do both value of `labels` and value of `score` shere need to be considered indexed by GT indices. just like this: `image_sorted_scores[''.join([chr(int(c)) for c in targets[i]['org_image_id']])] = {'labels':targets[i]['labels'][indices[i][1]].cpu().numpy(),"scores": prob[i,indices[i][0],targets[i]['labels'][indices[i][1]]].detach().cpu().numpy()}`

**Comment by FT115** [comment]
> Hi @FT115,
> 
> I look into this, but I think that the lines are different as in line 334, you want the target index.
> 
> Meanwhile in line 585, we are using the _source_ index, as we want the **score** of that prediction.
> 
> The `prob` array is indexed by the model predictions, rather than the GT labels. Unlike this, in line 334, `t["labels"]` is indexed by the GT indices. This is why in one, we want the source index and in the other the GT index. I will, however, run the code in debug to make sure this is the case.
> 
> Best, Orr

Thanks for your prompt reply, but it seems to me that here `prob` is used as value in dictionary, and another value is the GT label. So do both value of `labels` and value of `scores` here need to be considered indexed by GT indices. just like this: `image_sorted_scores[''.join([chr(int(c)) for c in targets[i]['org_image_id']])] = {'labels':targets[i]['labels'][indices[i][1]].cpu().numpy(),"scores": prob[i,indices[i][0],targets[i]['labels'][indices[i][1]]].detach().cpu().numpy()}`

**Comment by orrzohar** [comment]
Hi @FT115,

Looking at the code you sent, I where you are coming from. The `for` loop iterates over the images in the batch, and the dict should be a list of all the predictions:

```code
{
  image_id = {
    'labels'=[,...,],
    'scores'=[,...,],
  }
}
```

Scores should be a list of the target values. `prob` is BxPxC, so we need to, for image i, select the right elements from PxC. The main issue is that in:

https://github.com/orrzohar/PROB/blob/28afbc1d7f5bdfb9a384ce09c870fcd829adc1b2/models/prob_deformable_detr.py#L334

We needed one-hot vectors in the shape of the _predictions_ - i.e., for every prediction, we needed a one-hot vector denoting the matched target. However, in line:

https://github.com/orrzohar/PROB/blob/28afbc1d7f5bdfb9a384ce09c870fcd829adc1b2/models/prob_deformable_detr.py#L585

I am ordering the predictions in the order of the targets, and not vise-versa. We could, alternatively, do what you did, and order the lists w.r.t. the targets, however, I believe the same predictions would be selected just in different order (note that the additional index you added via the `indices` for the `'labels'` key will just re-order the target w.r.t. the predictions, and if you do that you will need to change `targets[i]['labels']-> indices[i][1]`

I will make sure this is the case. I do think your version is more intuitive, and will make sure my intuitions are correct. If they are, the two representations are equivalent, if I am wrong, then there is indeed a bug. I will update when I verify this. 
Best,
Orr

**Comment by FT115** [comment]
@orrzohar
Thank you for your reply, I understand what you are thinking.
